plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/AbstractDeltaLakePageSink.java

```diff
@@ -0,0 +1,566 @@
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.trinoVersion = requireNonNull(trinoVersion, "trinoVersion is null");
+        this.targetMaxFileSize = DeltaLakeSessionProperties.getTargetMaxFileSize(session);
+    }
+
+    protected abstract void processSynthesizedColumn(DeltaLakeColumnHandle column);
+
+    protected abstract void addSpecialColumns(
+            List<DeltaLakeColumnHandle> inputColumns,
+            ImmutableList.Builder<DeltaLakeColumnHandle> dataColumnHandles,
+            ImmutableList.Builder<Integer> dataColumnsInputIndex,
+            ImmutableList.Builder<String> dataColumnNames,
+            ImmutableList.Builder<Type> dataColumnTypes);
+
+    protected abstract String getPathPrefix();
+
+    protected abstract DataFileType getDataFileType();
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return writtenBytes;
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        return memoryUsage;
+    }
+
+    @Override
+    public long getValidationCpuNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public CompletableFuture<Collection<Slice>> finish()
+    {
+        for (int writerIndex = 0; writerIndex < writers.size(); writerIndex++) {
+            closeWriter(writerIndex);
+        }
+        writers.clear();
+
+        List<DataFileInfo> dataFilesInfo = dataFileInfos.build();
+        Collection<Slice> result = dataFilesInfo.stream()
+                .map(dataFileInfo -> wrappedBuffer(dataFileInfoCodec.toJsonBytes(dataFileInfo)))
+                .collect(toImmutableList());
+        return MoreFutures.toCompletableFuture(Futures.immediateFuture(result));
+    }
+
+    @Override
+    public void abort()
+    {
+        List<Closeable> rollbackActions = Streams.concat(
+                        writers.stream()
+                                // writers can contain nulls if an exception is thrown when doAppend expands the writer list
+                                .filter(Objects::nonNull)
+                                .map(writer -> writer::rollback),
+                        closedWriterRollbackActions.stream())
+                .collect(toImmutableList());
+        RuntimeException rollbackException = null;
+        for (Closeable rollbackAction : rollbackActions) {
+            try {
+                rollbackAction.close();
+            }
+            catch (Throwable t) {
+                if (rollbackException == null) {
+                    rollbackException = new TrinoException(DELTA_LAKE_BAD_WRITE, "Error rolling back write to Delta Lake");
+                }
+                rollbackException.addSuppressed(t);
+            }
+        }
+        if (rollbackException != null) {
+            throw rollbackException;
+        }
+    }
+
+    @Override
+    public CompletableFuture<?> appendPage(Page page)
+    {
+        if (page.getPositionCount() == 0) {
+            return NOT_BLOCKED;
+        }
```

Comment:
Same as above + I don't think it is redundant - It is true this method will return NOT_BLOCKED anyway but it will first try to call `writePage(page)`